stdpopsim/slim_engine.py

```diff
@@ -948,17 +948,54 @@ def fix_time(event):
     condition_on_allele_frequency = []
     op_id = stdpopsim.ext.ConditionOnAlleleFrequency.op_id
     slim_mutation_ids = []
-    for x in _dfe_to_mtypes(contig).values():
+    slim_mutation_dfe = []
+    for i, x in _dfe_to_mtypes(contig).items():
         slim_mutation_ids.extend([u[0] for u in x])
+        slim_mutation_dfe.extend([i for u in x])
     for ee in extended_events:
+        if hasattr(ee, "mutation_id"):
+            mut_id = getattr(ee, "mutation_id")
+            cls_name = ee.__class__.__name__
+            mtype_idx = [
+                j
+                for j, i in enumerate(slim_mutation_dfe)
+                if contig.dfe_list[i].id == mut_id
+            ]
+            n_mtypes = len(mtype_idx)
+            # TODO: are these checks too restrictive? For example, these will
+            # prevent using DrawMutation with an arbitrary MutationType in an
+            # arbitary DFE.
+            if n_mtypes != 1:
+                raise ValueError(
+                    f"{cls_name} event with mutation_id '{mut_id}' must "
+                    f"map to a single MutationType instance on {contig}, "
+                    f"but has {n_mtypes} associated MutationType instances."
+                )
+            else:
+                mtype_idx = mtype_idx[0]
+                setattr(ee, "mutation_type_id", slim_mutation_ids[mtype_idx])
+                dfe_idx = slim_mutation_dfe[mtype_idx]
+                dfe_intervals = contig.interval_list[dfe_idx]
+                if dfe_intervals.shape[0] > 0:
+                    raise ValueError(
+                        f"{cls_name} event with mutation_id '{mut_id}' is "
+                        f"a single site mutation, but is associated with "
+                        f"interval(s) {dfe_intervals}."
+                    )
+
         if hasattr(ee, "mutation_type_id"):
+            # TODO: this is a bit redundant now that ee.mutation_type
+            # replaces ee.mutation_type_id in the API, but should
+            # allow setattr(ee, "mutation_type_id", <int>) to work for
+            # arbitrary DFEs, if ee.mutation_type is None.
```

Comment:
Not sure I'm following - are you saying that leaving this bit of code in here would allow people do do the undocumented workflow
```
ee = ExtendedEvent(mutation_type=None, ...)
setattr(ee, "mutation_type_id", 1)
slim_makescript( ..., extended_events=[ee])
```
... ?

Reply:
So, I guess I think this code block should be part of the previous `if hasattr(ee, "mutation_id"):` code block. (We shouldn't be supporting that undocumented workflow).

Reply:
I was thinking in terms of pickled extended events created using the old `mutation_type_id` API being passed to the simulation engine. But you're right, if it's undocumented it should go. I could add a check that `mutation_type_id` is unset a bit earlier in that code block.

Reply:
We don't need to worry about this (or even backwards compatibility), as this hasn't been part of the released API.

Reply:
And, good thought, I really wouldn't worry about pickled old-style extended events. Anyone who's doing that is on their own here (and can deal with it).